contributors/design-proposals/local-storage-overview.md

```diff
@@ -0,0 +1,581 @@
+
+Since local PVs are only accessible from specific nodes, a new PV-node association will be used by the scheduler to place pods.  The association can be generalized to support any type of PV, not just local PVs.  This allows for any volume plugin to take advantage of this behavior.
+
+# User Workflows
+
+### Alice manages a deployment and requires “Guaranteed” ephemeral storage
+
+1. Kubelet running across all nodes will identify primary partition and expose capacity and allocatable for the primary partitions.  This allows primary partitions' storage capacity to be considered as a first class resource when scheduling.
+
+    ```yaml
+    apiVersion: v1
+    kind: Node
+    metadata:
+      name: foo
+    status:
+      capacity:
+        storage.kubernetes.io/overlay: 100Gi
+        storage.kubernetes.io/scratch: 100Gi
+      allocatable:
+        storage.kubernetes.io/overlay: 100Gi
+        storage.kubernetes.io/scratch: 90Gi
+    ```
+
+2. Alice adds new storage resource requirements to her pod, specifying limits for the container's writeable and overlay layers, and emptyDir volumes.
+
+    ```yaml
+    apiVersion: v1
+    kind: pod
+    metadata:
+     name: foo
+    spec:
+     containers:
+     - name: fooc
+       resources:
+         limits:
+           storage.kubernetes.io/logs: 500Mi
+           storage.kubernetes.io/overlay: 1Gi
+       volumeMounts:
+       - name: myEmptyDir
+         mountPath: /mnt/data
+     volumes:
+     - name: myEmptyDir
+       emptyDir:
+	     size: 20Gi
+    ```
+
+3. Alice’s pod “foo” is Guaranteed a total of “21.5Gi” of local storage. The container “fooc” in her pod cannot consume more than 1Gi for writable layer and 500Mi for logs, and “myEmptyDir” volume cannot consume more than 20Gi.
+4. For the pod resources, `storage.kubernetes.io/logs` resource is meant for logs. `storage.kubernetes.io/overlay` is meant for writable layer.
+5. `storage.kubernetes.io/logs` is satisfied by `storage.kubernetes.io/scratch`.
+6. `storage.kubernetes.io/overlay` resource can be satisfied by `storage.kubernetes.io/overlay` if exposed by nodes or by `storage.kubernetes.io/scratch` otherwise. The scheduler follows this policy to find an appropriate node which can satisfy the storage resource requirements of the pod.
+7. EmptyDir.size is both a request and limit that is satisfied by `storage.kubernetes.io/scratch`. 
+8. Kubelet will rotate logs to keep scratch space usage of “fooc” under 500Mi
+9. Kubelet will track the usage of pods across logs and overlay filesystem and restart the container if it's total usage exceeds it's storage limits. If usage on `EmptyDir` volume exceeds its `limit`, then the pod will be evicted by the kubelet. By performing soft limiting, users will be able to easily identify pods that run out of storage.
```

Comment:
> Is it an optimization to just nuke emptyDirs and restart locally

This proposal states that whenever a pod exceeds it's emptydir limits it will be `Evicted`. Evicted pods are never restarted. It's a terminal phase. I'm concerned with restarting locally before reclaiming disk space. Since reclamation takes a long time, I don't know if it makes sense to require local restarts. I'm hoping that vertical auto-scaling will reduce the probability of storage related evictions.

> Is it an optimization to just nuke emptyDirs and restart locally

CRI will be updated to require writable layer usage metrics. If a runtime cannot expose that metric, then it cannot provide complete isolation. Kubelet will not reject the pod. Instead, it will not be able to enforce storage. Imagine a buggy runtime that does exposes incorrect writable layer usage. That scenario is no different from a runtime that does not expose writable layer usage at all from kubelet's perspective.

> . will we actually allow pods to exceed their limits, or prevent them from exceeding their limits for emptyDir?

To begin with yes, pods will be able to exceed limits which might result in allocatable level evictions. Until linux project quota is available this will be the case. For emptyDir we do have the option of use gid based linux quota but the initial implementation will not use that. Baby steps towards a robust system.